### PR TITLE
Add service-proxy at agent-side, use default MPSR to fix the security issue.

### PR DESCRIPTION
**Pull request by xuezhaojun**
The related user-story: https://github.com/stolostron/backlog/issues/26546
The related epic: https://github.com/stolostron/backlog/issues/25894

### CODE REVIEW

**Comment by xuezhaojun** [comment]
Pass the test in local env, it's ready for review now. Continue to work on adding more test-cases and e2e.

File: pkg/addon/cmd.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+		klog.Errorf("failed to new addon manager %v", err)
+		return err
+	}
+
+	registrationOption := NewRegistrationOption(
+		kubeConfig,
+		AddonName)
+
+	af := addonfactory.NewAgentAddonFactory(AddonName, FS, "manifests/charts/service-proxy").
+		WithGetValuesFuncs(GetValues(kubeClient, certificatesNamespace, agentImage), addonfactory.GetValuesFromAddonAnnotation).WithAgentRegistrationOption(registrationOption)
```
Comment (skeeey) [inline pkg/addon/cmd.go:63]: TODO support addondeploymentconfig

File: pkg/addon/manifests/charts/service-proxy/Chart.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+apiVersion: v2
+name: cluster-proxy-service-proxy
+description: A Helm chart for Cluster-Proxy Addon - Service-Proxy (Agent Only)
+type: application
+version: 0.2.2
+appVersion: 1.0.0
```
Comment (skeeey) [inline pkg/addon/manifests/charts/service-proxy/Chart.yaml:6]: new line

File: pkg/addon/manifests/charts/service-proxy/templates/service-proxy-deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            mountPath: /apiserver-ca
+          - name: ocpservice-ca
+            mountPath: /ocpservice-ca
+          - name: service-proxy-server-cert
+            mountPath: /server-cert
+            readOnly: true
+      {{- if .Values.global.imagePullSecret }}
+      imagePullSecrets:
+        - name: "{{ .Values.global.imagePullSecret }}"
+      {{- end }}
```
Comment (skeeey) [inline pkg/addon/manifests/charts/service-proxy/templates/service-proxy-deployment.yaml:66]: new line

File: pkg/addon/manifests/charts/service-proxy/templates/serviceaccount.yaml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+kind: ServiceAccount
+apiVersion: v1
+metadata:
+  name: cluster-proxy-service-proxy-sa
+  namespace: {{ .Values.addonInstallNamespace }}
+  labels:
+    app: {{ .Values.addonName }}
```
Comment (skeeey) [inline pkg/addon/manifests/charts/service-proxy/templates/serviceaccount.yaml:7]: new line

File: pkg/addon/manifests/charts/service-proxy/values.yaml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    memory: 128Mi
+  limits:
+    memory: 512Mi
+
+global:
+  imagePullPolicy: IfNotPresent
+  imagePullSecret: open-cluster-management-image-pull-credentials
+  imageOverrides:
+    cluster_proxy_addon: quay.io/open-cluster-management/cluster-proxy-addon:latest
+  nodeSelector: {}
```
Comment (skeeey) [inline pkg/addon/manifests/charts/service-proxy/values.yaml:30]: new line

File: pkg/utils/utils.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+		Service:   "kubernetes",
+		Namespace: "default",
+		Port:      "443",
+		Path:      kubeAPIPath,
+	}, nil
+}
+
+func GetTargetServiceConfig(requestURL string) (ts TargetServiceConfig, err error) {
+	urlparams := strings.Split(requestURL, "/")
+
```
Comment (skeeey) [inline pkg/utils/utils.go:52]: check the urlparams length firstly?

File: pkg/serviceproxy/service_proxy.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+	proxy.Transport = &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		DialContext: (&net.Dialer{
+			Timeout:   30 * time.Second,
+			KeepAlive: 30 * time.Second,
+		}).DialContext,
+		MaxIdleConns:          100,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
```
Comment (skeeey) [inline pkg/serviceproxy/service_proxy.go:114]: TODO may config these times outside

File: pkg/addon/cmd.go (hunk 1 above)
Comment (xuezhaojun) [reply pkg/addon/cmd.go:63]: TODO added, will support `addondeploymentconfig` in other PR in the next Sprint.

File: pkg/serviceproxy/service_proxy.go (hunk 7 above)
Comment (xuezhaojun) [reply pkg/serviceproxy/service_proxy.go:114]: Fixed, now they are configurable via flags.

File: pkg/utils/utils.go (hunk 6 above)
Comment (xuezhaojun) [reply pkg/utils/utils.go:52]: Fixed.

**Comment by xuezhaojun** [comment]
/hold

Need to change addon-framework to a controller to avoid another addon CR.

File: pkg/controllers/deploycontroller.go [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+	if evt.ObjectNew.GetName() != constant.AddonName {
+		return
+	}
+	q.Add(reconcile.Request{NamespacedName: types.NamespacedName{
+		Name:      evt.ObjectNew.GetName(),
+		Namespace: evt.ObjectNew.GetNamespace(),
+	}})
+}
+
+func (h *managedcluteraddonHandler) Delete(evt event.DeleteEvent, q workqueue.RateLimitingInterface) {
```
Comment (skeeey) [inline pkg/controllers/deploycontroller.go:233]: Need we consider add-on deletion?

File: pkg/controllers/deploycontroller.go [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+	managedClusterAddonName := request.Name
+	managedClusterName := request.Namespace
+	if len(managedClusterName) == 0 || len(managedClusterAddonName) == 0 {
+		klog.Infof("Skip reconcile: managedClusterAddonName or managedClusterName is empty")
+		return reconcile.Result{}, nil
+	}
+
+	// get managedCluster
+	managedCluster, err := r.clusterLister.Get(managedClusterName)
+	if err != nil {
```
Comment (skeeey) [inline pkg/controllers/deploycontroller.go:93]: if not found, ignore?

File: pkg/controllers/deploycontroller.go [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+		klog.Errorf("Failed to get managedCluster %q: %v", managedClusterName, err)
+		return reconcile.Result{}, err
+	}
+
+	// get managedClusterAddon
+	managedClusterAddon, err := r.addonLister.ManagedClusterAddOns(managedClusterName).Get(managedClusterAddonName)
+	if err != nil {
```
Comment (skeeey) [inline pkg/controllers/deploycontroller.go:100]: ignore not found?

File: pkg/controllers/deploycontroller.go (hunk 9 above)
Comment (skeeey) [reply pkg/controllers/deploycontroller.go:93]: if managed cluster is deleting, ignore?

File: pkg/controllers/deploycontroller.go [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+
+	// get override image name
+	overrideImageName, err := getOverrideImageName(managedCluster, agentImage)
+	if err != nil {
+		klog.Errorf("Failed to get override image name: %v", err)
+		return reconcile.Result{}, err
+	}
+
+	// get nodeSelector
+	nodeSelector, err := getNodeSelector(managedCluster)
```
Comment (skeeey) [inline pkg/controllers/deploycontroller.go:120]: if non local-cluster, how do we handle this?

File: pkg/controllers/deploycontroller.go (hunk 8 above)
Comment (skeeey) [reply pkg/controllers/deploycontroller.go:233]: ok, you have an owner on the manifest work

File: pkg/controllers/deploycontroller.go (hunk 11 above)
Comment (xuezhaojun) [reply pkg/controllers/deploycontroller.go:120]: If it's not a local-cluter, it wil be empty.
I think we need to handle addondeployment config also in this release with another PR to solve this.

File: pkg/controllers/deploycontroller.go (hunk 9 above)
Comment (xuezhaojun) [reply pkg/controllers/deploycontroller.go:93]: Yes, here we skip "not found and deleting", this controller only acts as a reader.

File: pkg/controllers/deploycontroller.go (hunk 10 above)
Comment (xuezhaojun) [reply pkg/controllers/deploycontroller.go:100]: Explained in the above thread.

File: pkg/controllers/deploycontroller.go (hunk 11 above)
Comment (xuezhaojun) [reply pkg/controllers/deploycontroller.go:136]: TODO added.

File: pkg/controllers/deploycontroller.go (hunk 10 above)
Comment (xuezhaojun) [reply pkg/controllers/deploycontroller.go:108]: Fixed.

File: pkg/controllers/deploycontroller.go (hunk 9 above)
Comment (xuezhaojun) [reply pkg/controllers/deploycontroller.go:93]: Fixed.

**Pull request merged**